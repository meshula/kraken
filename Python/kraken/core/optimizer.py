"""Kraken - traverser module.

Classes:
Traverser - Base Traverser.

"""

from kraken.core.objects.scene_item import SceneItem
from kraken.core.objects.object_3d import Object3D
from kraken.core.objects.components.component import Component
from kraken.core.objects.attributes.attribute_group import AttributeGroup
from kraken.core.objects.constraints.constraint import Constraint
from kraken.core.objects.operators.operator import Operator


class Optimizer(object):
    """Kraken base traverser for any scene item.

    The optimizer iterates through all scene items and collapses resundancies by deleting items and rewiring connections.

    """

    def __init__(self, name='Optimizer'):
        pass

    def optimize(self):
        pass
