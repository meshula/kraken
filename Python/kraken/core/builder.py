"""Kraken - builders module.

Classes:
Builder -- Base builder object to build objects in DCC.

"""


from kraken.core.kraken_system import KrakenSystem
from kraken.core.configs.config import Config
from kraken.core.profiler import Profiler

from kraken.core.objects.components.component import Component
from kraken.core.objects.constraints.pose_constraint import PoseConstraint


class Builder(object):
    """Builder object for building objects in DCC's. Sub-class per DCC in a
    plugin."""


    def __init__(self, debugMode = False):
        super(Builder, self).__init__()
        self._buildElements = []

        self.config = Config.getInstance()

        self._debugMode = debugMode


    # ====================
    # Object registration
    # ====================
    def _registerSceneItemPair(self, kSceneItem, dccSceneItem):
        """Registers a pairing between the kraken scene item and the dcc scene item
        for querying later.

        Args:
            kSceneItem (object): kraken scene item that you want to pair.
            dccSceneItem (object): dcc scene item that you want to pair.

        Returns:
            bool: True if successful.

        """

        pairing = {
                   'src': kSceneItem,
                   'tgt': dccSceneItem
                  }

        self._buildElements.append(pairing)

        return True

    def deleteBuildElements(self):
        """
        Clear out all dcc built elements from the scene if exist
        """

        return None


    def getDCCSceneItem(self, kSceneItem):
        """Given a kSceneItem, returns the built dcc scene item.

        Args:
            kSceneItem (object): kSceneItem to base the search.

        Returns:
            object: The DCC Scene Item that corresponds to the given scene item

        """

        for builtElement in self._buildElements:

            if builtElement['src'] == kSceneItem:
                return builtElement['tgt']


        return None


    # ========================
    # SceneItem Build Methods
    # ========================
    def buildContainer(self, kSceneItem, buildName):
        """Builds a container / namespace object.

        Args:
            kSceneItem (object): kSceneItem that represents a container to be built.
            buildName (string): The name to use on the built object.

        Returns:
            object: DCC Scene Item that is created.

        """

        return None


    def buildLayer(self, kSceneItem, buildName):
        """Builds a layer object.

        Args:
            kSceneItem (object): kSceneItem that represents a layer to be built.
            buildName (string): The name to use on the built object.

        Returns:
            object: DCC Scene Item that is created.

        """

        return None


    def buildHierarchyGroup(self, kSceneItem, buildName):
        """Builds a hierarchy group object.

        Args:
            kSceneItem (object): kSceneItem that represents a group to be built.
            buildName (string): The name to use on the built object.

        Returns:
            object: DCC Scene Item that is created.

        """

        return None


    def buildGroup(self, kSceneItem, buildName):
        """Builds a group object.

        Args:
            kSceneItem (object): kSceneItem that represents a group to be built.
            buildName (string): The name to use on the built object.

        Returns:
            object: DCC Scene Item that is created.

        """

        return None


    def buildJoint(self, kSceneItem, buildName):
        """Builds a joint object.

        Args:
            kSceneItem (object): kSceneItem that represents a joint to be built.
            buildName (string): The name to use on the built object.

        Returns:
            object: DCC Scene Item that is created.

        """

        return None


    def buildLocator(self, kSceneItem, buildName):
        """Builds a locator / null object.

        Args:
            kSceneItem (object): kSceneItem that represents a locator / null to be built.
            buildName (string): The name to use on the built object.

        Returns:
            object: DCC Scene Item that is created.

        """

        return None


    def buildCurve(self, kSceneItem, buildName):
        """Builds a Curve object.

        Args:
            kSceneItem (object): kSceneItem that represents a curve to be built.
            buildName (string): The name to use on the built object.

        Returns:
            object: DCC Scene Item that is created.

        """

        return None


    def buildControl(self, kSceneItem, buildName):
        """Builds a Control object.

        Args:
            kSceneItem (object): kSceneItem that represents a control to be built.
            buildName (string): The name to use on the built object.

        Returns:
            object: DCC Scene Item that is created.

        """

        return None


    # ========================
    # Attribute Build Methods
    # ========================
    def buildBoolAttribute(self, kAttribute):
        """Builds a Bool attribute.

        Args:
            kAttribute (object): kAttribute that represents a string attribute to be built.

        Returns:
            object: True if successful.

        """

        return True


    def buildScalarAttribute(self, kAttribute):
        """Builds a Float attribute.

        Args:
            kAttribute (object): kAttribute that represents a string attribute to be built.

        Returns:
            bool: True if successful.

        """

        return True


    def buildIntegerAttribute(self, kAttribute):
        """Builds a Integer attribute.

        Args:
            kAttribute (object): kAttribute that represents a string attribute to be built.

        Returns:
            bool: True if successful.

        """

        return True


    def buildStringAttribute(self, kAttribute):
        """Builds a String attribute.

        Args:
            kAttribute (object): kAttribute that represents a string attribute to be built.

        Returns:
            bool: True if successful.

        """

        return True


    def buildAttributeGroup(self, kAttributeGroup):
        """Builds attribute groups on the DCC object.

        Args:
            kAttributeGroup (SceneObject): kraken object to build the attribute group on.

        Returns:
            bool: True if successful.

        """

        return True


    def connectAttribute(self, kAttribute):
        """Connects the driver attribute to this one.

        Args:
            kAttribute (object): attribute to connect.

        Returns:
            bool: True if successful.

        """

        return True


    # =========================
    # Constraint Build Methods
    # =========================
    def buildOrientationConstraint(self, kConstraint):
        """Builds an orientation constraint represented by the kConstraint.

        Args:
            kConstraint (object): kraken constraint object to build.

        Returns:
            object: DCC Scene Item that was created.

        """

        constraineeDCCSceneItem = self.getDCCSceneItem(kConstraint.getConstrainee())
        dccSceneItem = None # Add constraint object here.
        self._registerSceneItemPair(kConstraint, dccSceneItem)

        return dccSceneItem


    def buildPoseConstraint(self, kConstraint):
        """Builds an pose constraint represented by the kConstraint.

        Args:
            kConstraint (object): kraken constraint object to build.

        Returns:
            bool: True if successful.

        """

        constraineeDCCSceneItem = self.getDCCSceneItem(kConstraint.getConstrainee())
        dccSceneItem = None # Add constraint object here.
        self._registerSceneItemPair(kConstraint, dccSceneItem)

        return dccSceneItem


    def buildPositionConstraint(self, kConstraint):
        """Builds an position constraint represented by the kConstraint.

        Args:
            kConstraint (object): kraken constraint object to build.

        Returns:
            bool: True if successful.

        """

        constraineeDCCSceneItem = self.getDCCSceneItem(kConstraint.getConstrainee())
        dccSceneItem = None # Add constraint object here.
        self._registerSceneItemPair(kConstraint, dccSceneItem)

        return dccSceneItem


    def buildScaleConstraint(self, kConstraint):
        """Builds an scale constraint represented by the kConstraint.

        Args:
            kConstraint (object): kraken constraint object to build.

        Returns:
            bool: True if successful.

        """

        constraineeDCCSceneItem = self.getDCCSceneItem(kConstraint.getConstrainee())
        dccSceneItem = None # Add constraint object here.
        self._registerSceneItemPair(kConstraint, dccSceneItem)

        return dccSceneItem


    # ========================
    # Component Build Methods
    # ========================
    def buildXfoConnection(self, componentInput):
        """Builds the constraint between the target and connection target.

        Args:
            componentInput (object): kraken component input to build connections for.

        Returns:
            bool: True if successful.

        """

        if componentInput.isConnected() is False:
            return False

        connection = componentInput.getConnection()
        connectionTarget = connection.getTarget()
        inputTarget = componentInput.getTarget()

        if connection.getDataType().endswith('[]'):
            if componentInput.getIndex() > len(connection.getTarget()) - 1:

                inputParent = componentInput.getParent()
                inputParentDecoration = inputParent.getNameDecoration()
                fullInputName = inputParent.getName() + inputParentDecoration + "." + componentInput.getName()

                raise Exception(fullInputName + " index ("
                                + str(componentInput.getIndex()) + ") is out of range ("
                                + str(len(connection.getTarget()) - 1) + ")!")

            connectionTarget = connection.getTarget()[componentInput.getIndex()]
        else:
            connectionTarget = connection.getTarget()

        # There should be no offset between an output xfo from one component and the connected input of another
        # If connected, they should be exactly the same.
        constraint = inputTarget.constrainTo(connectionTarget, maintainOffset=False, addToConstraintList=False)

        dccSceneItem = self.buildPoseConstraint(constraint)
        self._registerSceneItemPair(componentInput, dccSceneItem)

        return True


    def buildAttributeConnection(self, componentInput):
        """Builds the link between the target and connection target.

        Args:
            componentInput (object): kraken connection to build.

        Returns:
            bool: True if successful.

        """

        # Implemented in DCC Plugins.

        return True


    # =========================
    # Operator Builder Methods
    # =========================
    def buildKLOperator(self, kKLOperator):
        """Builds Splice Operators on the components.

        Args:
            kOperator (object): kraken operator that represents a KL operator.

        Returns:
            bool: True if successful.

        """

        return True

    def buildCanvasOperator(self, kOperator):
        """Builds Splice Operators on the components.

        Args:
            kOperator (object): kraken operator that represents a Canvas operator.

        Returns:
            bool: True if successful.

        """

        return True


    # =====================
    # Build Object Methods
    # =====================
    def buildAttributes(self, kObject):
        """Builds attributes on the DCC object.

        Args:
            kObject (SceneObject): kraken object to build attributes for.

        Returns:
            bool: True if successful.

        """

        for i in xrange(kObject.getNumAttributeGroups()):
            attributeGroup = kObject.getAttributeGroupByIndex(i)
            self.buildAttributeGroup(attributeGroup)

        return True


    def buildHierarchy(self, kObject, component=None):
        """Builds the hierarchy for the supplied kObject.

        Args:
            kObject (object): kraken object to build.
            component (Component): component that this object belongs to.

        Returns:
            object: DCC object that was created.

        """


        dccSceneItem = None

        buildName = kObject.getBuildName()

        if self._debugMode:
            print "building:" + kObject.getPath() + " as:" + buildName

        # Build Object
        if kObject.isTypeOf("Container"):
            dccSceneItem = self.buildContainer(kObject, buildName)

        elif kObject.isTypeOf("Layer"):
            dccSceneItem = self.buildLayer(kObject, buildName)

        elif kObject.isTypeOf("Component"):
            pass

        elif kObject.isTypeOf("ComponentGroup"):
            dccSceneItem = self.buildGroup(kObject, buildName)
            component = kObject

        elif kObject.isTypeOf("HierarchyGroup"):
            dccSceneItem = self.buildHierarchyGroup(kObject, buildName)

        elif kObject.isTypeOf("CtrlSpace"):
            dccSceneItem = self.buildGroup(kObject, buildName)

        elif kObject.isTypeOf("Transform"):
            dccSceneItem = self.buildGroup(kObject, buildName)

        elif kObject.isTypeOf("Locator"):
            dccSceneItem = self.buildLocator(kObject, buildName)

        elif kObject.isTypeOf("Joint"):
            dccSceneItem = self.buildJoint(kObject, buildName)

        elif kObject.isTypeOf("Control"):
            dccSceneItem = self.buildControl(kObject, buildName)

        elif kObject.isTypeOf("Curve"):
            dccSceneItem = self.buildCurve(kObject, buildName)

        # Important Note: The order of these tests is important.
        # New classes should be added above the classes they are derrived from.
        # No new types should be added below SceneItem here.
        elif kObject.isTypeOf("SceneItem"):
            dccSceneItem = self.buildLocator(kObject, buildName)

        else:
            raise NotImplementedError(kObject.getName() + ' has an unsupported type: ' + str(type(kObject)))

        if dccSceneItem is not None:
            self.buildAttributes(kObject)
            self.setTransform(kObject)
            self.lockParameters(kObject)
            self.setVisibility(kObject)
            self.setObjectColor(kObject)

        # Build children
        for i in xrange(kObject.getNumChildren()):
            child = kObject.getChildByIndex(i)
            self.buildHierarchy(child, component)

        return dccSceneItem


    def buildConstraints(self, kObject):
        """Builds constraints for the supplied kObject.

        Args:
            kObject (object): kraken object to create constraints for.

        Returns:
            bool: True if successful.

        """

        dccSceneItem = None
        for i in xrange(kObject.getNumConstraints()):

            constraint = kObject.getConstraintByIndex(i)

            # Build Object
            if constraint.isTypeOf("OrientationConstraint"):
                dccSceneItem = self.buildOrientationConstraint(constraint)

            elif constraint.isTypeOf("PoseConstraint"):
                dccSceneItem = self.buildPoseConstraint(constraint)

            elif constraint.isTypeOf("PositionConstraint"):
                dccSceneItem = self.buildPositionConstraint(constraint)

            elif constraint.isTypeOf("ScaleConstraint"):
                dccSceneItem = self.buildScaleConstraint(constraint)

            else:
                raise NotImplementedError(constraint.getName() + ' has an unsupported type: ' + str(type(constraint)))

        # Build children
        for i in xrange(kObject.getNumChildren()):
            child = kObject.getChildByIndex(i)
            self.buildConstraints(child)

        return True


    def buildInputConnections(self, kObject):
        """Builds the connections between the component inputs of each
        component.

        Only input connections are built otherwise duplicate constraints / expressions
        would be created.

        Args:
            kObject (object): kraken object to create connections for.

        Returns:
            bool: True if successful.

        """

        if kObject.isTypeOf('Component'):

            # Build input connections
            for i in xrange(kObject.getNumInputs()):

                componentInput = kObject.getInputByIndex(i)
                if componentInput.getTarget() is None or componentInput.getConnection() is None:
                    continue

                if self._debugMode:
                    print "buildConnection:" + componentInput.getName()

                if componentInput.getDataType().startswith('Xfo'):
                    self.buildXfoConnection(componentInput)

                elif componentInput.getDataType().startswith(('Boolean', 'Float', 'Integer', 'String')):
                    self.buildAttributeConnection(componentInput)

        # Build connections for children.
        for i in xrange(kObject.getNumChildren()):
            child = kObject.getChildByIndex(i)
            self.buildInputConnections(child)

        return True


    def buildAttrConnections(self, kObject):
        """Builds the connections between the component inputs and outputs of each
        component.

        Args:
            kObject (object): kraken object to create connections for.

        Returns:
            bool: True if successful.

        """

        # Build input connections
        for i in xrange(kObject.getNumAttributeGroups()):
            attributeGroup = kObject.getAttributeGroupByIndex(i)
            for y in xrange(attributeGroup.getNumAttributes()):
                attribute = attributeGroup.getAttributeByIndex(y)
                self.connectAttribute(attribute)

        # Build connections for children.
        for i in xrange(kObject.getNumChildren()):
            child = kObject.getChildByIndex(i)
            self.buildAttrConnections(child)

        return True


    def buildOperators(self, kObject):
        """Build operators in the hierarchy.

        Args:
            kObject (object): kraken object to create operators for.

        Returns:
            bool: True if successful.

        """

        if kObject.isTypeOf('Component'):

            # Build operators
            for i in xrange(kObject.getNumOperators()):
                operator = kObject.getOperatorByIndex(i)

                if operator.isTypeOf('KLOperator'):
                    self.buildKLOperator(operator)
                elif operator.isTypeOf('CanvasOperator'):
                    self.buildCanvasOperator(operator)
                else:
                    raise NotImplementedError(operator.getName() + ' has an unsupported type: ' + str(type(operator)))

        # Build connections for children.
        for i in xrange(kObject.getNumChildren()):
            child = kObject.getChildByIndex(i)
            self.buildOperators(child)

        return True


    # ==================
    # Parameter Methods
    # ==================
    def lockParameters(self, kSceneItem):
        """Locks flagged SRT parameters.

        Args:
            kSceneItem (object): kraken object to lock the SRT parameters on.

        Returns:
            bool: True if successful.

        """

        return True


    # ===================
    # Visibility Methods
    # ===================
    def setVisibility(self, kSceneItem):
        """Sets the visibility of the object after its been created.

        Args:
            kSceneItem (object): kraken object to set the visibility on.

        Returns:
            bool: True if successful.

        """

        if hasattr(kSceneItem, 'getShapeVisibility') is False:
            return False

        if kSceneItem.getShapeVisibility() is False:
            pass

            # Re-implement in DCC builders.

        return True


    # ================
    # Display Methods
    # ================
    def getBuildColor(self, kSceneItem):
        """Gets the build color for the object.

        Args:
            kSceneItem (object): kraken object to get the color for.

        Returns:
            str: color to set on the object.

        """

        config = self.getConfig()
        colors = config.getColors()
        colorMap = config.getColorMap()
        typeNames = kSceneItem.getTypeHierarchyNames()
        component = kSceneItem.getComponent()
        objectColor = kSceneItem.getColor()

        buildColor = None
        if objectColor is not None:

            if objectColor not in colors.keys():
                raise ValueError("Invalid color '" + objectColor + "' set on: " + kSceneItem.getPath())

            buildColor = objectColor

        else:
            # Find the first color mapping that matches a type in the object hierarchy.
            for typeName in typeNames:
                if typeName in colorMap.keys():
                    if component is None:
                        buildColor = colorMap[typeName]['default']
                    else:
                        componentLocation = component.getLocation()
                        buildColor = colorMap[typeName][componentLocation]
                    break

        return buildColor


    def setObjectColor(self, kSceneItem):
        """Sets the color on the dccSceneItem.

        Args:
            kSceneItem (object): kraken object to set the color on.

        Returns:
            bool: True if successful.

        """

        return True


    # ==================
    # Transform Methods
    # ==================
    def setTransform(self, kSceneItem):
        """Translates the transform to Softimage transform.

        Args:
            kSceneItem (object): object to set the transform on.

        Returns:
            bool: True if successful.

        """

        dccSceneItem = self.getDCCSceneItem(kSceneItem)

        # Re-implement in DCC builders.

        return True


    # ===============
    # Config Methods
    # ===============
    def getConfig(self):
        """Gets the current config for the builder.

        Returns:
            object: config.

        """

        return self.config


    def setConfig(self, config):
        """Sets the builder's config.

        Args:
            config (Config): the config to use for this builder.

        Returns:
            bool: True if successful.

        """

        self.config = config

        return True


    # ==============
    # Build Methods
    # ==============
    def _preBuild(self, kSceneItem):
        """Protected Pre-Build method.

        Args:
            kSceneItem (object): kraken kSceneItem object to build.

        Returns:
            bool: True if successful.

        """

        return True


    def _build(self, kSceneItem):
        """Protected build method.

        Args:
            kSceneItem (object): kraken kSceneItem object to build.

        Returns:
            bool: True if successful.

        """

        self.buildHierarchy(kSceneItem, component=None)
        self.buildAttrConnections(kSceneItem)
        self.buildInputConnections(kSceneItem)
        self.buildOperators(kSceneItem)
        self.buildConstraints(kSceneItem)

        return True


    def build(self, kSceneItem):
        """Builds the supplied kSceneItem into a DCC representation.

        Args:
            kSceneItem (object): kraken kSceneItem object to build.

        Returns:
            object: The DCC scene item of the kSceneItem that was passed to the builder.

        """

        Profiler.getInstance().push("build:" + kSceneItem.getName())

        try:
            self._preBuild(kSceneItem)
            self._build(kSceneItem)

        finally:
            self._postBuild()

            # Clear Config when finished.
            self.config.clearInstance()

        Profiler.getInstance().pop()

        return self.getDCCSceneItem(kSceneItem)


    def _postBuild(self):
        """Protected Post-Build method.

        Returns:
            bool: True if successful.

        """

        return True

