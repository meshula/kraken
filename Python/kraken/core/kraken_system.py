"""KrakenSystem - objects.kraken_core module.

Classes:
KrakenSystem - Class for constructing the Fabric Engine Core client.

"""

import os
import sys
import json
import imp
import inspect
import importlib
from collections import OrderedDict
import traceback

import FabricEngine.Core

import kraken
from kraken.core.profiler import Profiler

krakenSystemModuleDir = os.path.join(os.path.dirname(os.path.realpath(__file__)))
krakenDir=os.path.abspath(os.path.join(krakenSystemModuleDir, '..', '..', '..'))
os.environ['KRAKEN_PATH']  = krakenDir

krakenExtsDir = os.path.join(krakenDir, 'KLExts')
if krakenExtsDir not in os.environ['FABRIC_EXTS_PATH']:
    os.environ['FABRIC_EXTS_PATH'] = krakenExtsDir + ';' + os.environ['FABRIC_EXTS_PATH']

canvasPresetsDir = os.path.join(krakenDir, 'CanvasPresets')
if 'FABRIC_DFG_PATH' in os.environ:
    if canvasPresetsDir not in os.environ['FABRIC_DFG_PATH']:
        os.environ['FABRIC_DFG_PATH'] = canvasPresetsDir + ';' + os.environ['FABRIC_DFG_PATH']
else:
    os.environ['FABRIC_DFG_PATH'] = canvasPresetsDir

class KrakenSystem(object):
    """The KrakenSystem is a singleton object used to provide an interface with
    the FabricEngine Core and RTVal system."""

    __instance = None

    def __init__(self):
        """Initializes the Kraken System object."""

        super(KrakenSystem, self).__init__()

        self.client = None
        self.typeDescs = None
        self.registeredTypes = None
        self.loadedExtensions = []

        self.registeredConfigs = OrderedDict()
        self.registeredComponents = OrderedDict()
        # self.moduleImportManager = ModuleImportManager()


    def loadCoreClient(self):
        """Loads the Fabric Engine Core Client"""

        if self.client == None:
            Profiler.getInstance().push("loadCoreClient")

            try:
                imp.find_module('cmds')
                host = 'Maya'
            except ImportError:
                try:
                    imp.find_module('sipyutils')
                    host = 'Softimage'
                except ImportError:
                    host = 'Python'

            if host == "Python":
                self.client = FabricEngine.Core.createClient()

            elif host == "Maya":
                contextID = cmds.fabricSplice('getClientContextID')
                if contextID == '':
                    cmds.fabricSplice('constructClient')
                    contextID = cmds.fabricSplice('getClientContextID')

                # Pull out the Splice client.
                self.client = FabricEngine.Core.createClient({"contextID": contextID})

            elif host == "Softimage":
                from win32com.client.dynamic import Dispatch
                si = Dispatch("XSI.Application").Application
                contextID = si.fabricSplice('getClientContextID')
                if contextID == '':
                    si.fabricSplice('constructClient')
                    contextID = si.fabricSplice('getClientContextID')

                # Pull out the Splice client.
                self.client = FabricEngine.Core.createClient({"contextID": contextID})

            self.loadExtension('Math')

            # krakenDir = os.environ['KRAKEN_PATH']
            # self.client.DFG.host.addPresetDir('', 'Kraken', os.path.join(krakenDir, 'CanvasPresets'))

            Profiler.getInstance().pop()


    def getCoreClient(self):
        """Returns the Fabric Engine Core Client owned by the KrakenSystem

        Returns:
            object: The Fabric Engine Core Client

        """

        if self.client is None:
            self.loadCoreClient()

        return self.client


    def loadExtension(self, extension):
        """Loads the given extension and updates the registeredTypes cache.

        Args:
            extension (str): The name of the extension to load.

        """

        if extension not in self.loadedExtensions:
            Profiler.getInstance().push("loadExtension:" + extension)
            self.client.loadExtension(extension)
            self.registeredTypes = self.client.RT.types
            self.typeDescs = self.client.RT.getRegisteredTypes()
            # Cache the loaded extension so that we aviod refreshing the typeDescs cache(costly)
            self.loadedExtensions.append(extension)
            Profiler.getInstance().pop()

    # ==============
    # RTVal Methods
    # ==============

    def constructRTVal(self, dataType, defaultValue=None):
        """Constructs a new RTVal using the given name and optional devault value.

        Args:
            dataType (str): The name of the data type to construct.
            defaultValue (value): The default value to use to initialize the RTVal

        Returns:
            object: The constructed RTval.

        """

        self.loadCoreClient()
        klType = getattr(self.registeredTypes, dataType)

        if defaultValue is not None:
            if hasattr(defaultValue, '_rtval'):
                return defaultValue._rtval

            typeDesc = self.typeDescs[dataType]
            if 'members' in typeDesc:
                try:
                    value = klType.create()
                except:
                    value = klType()
                for i in range(0, len(typeDesc['members'])):
                    memberName = typeDesc['members'][i]['name']
                    memberType = typeDesc['members'][i]['type']
                    if memberName in defaultValue:
                        setattr(value, memberName, self.constructRTVal(memberType, getattr(defaultValue, memberName)))
                return value

            else:
                return klType(defaultValue)
        else:
            try:
                return klType.create()
            except:
                try:
                    return klType()
                except Exception as e:
                    raise Exception("Error constructing RTVal:" + dataType)


    def rtVal(self, dataType, defaultValue=None):
        """Constructs a new RTVal using the given name and optional devault value.

        Args:
            dataType (str): The name of the data type to construct.
            defaultValue (value): The default value to use to initialize the RTVal

        Returns:
            object: The constructed RTval.

        """

        return self.constructRTVal(dataType, defaultValue)


    def isRTVal(self, value):
        """Returns true if the given value is an RTVal.

        Args:
            value (value): value to test.

        Returns:
            bool: True if successful.

        """

        return str(type(value)) == "<type 'PyRTValObject'>"


    def getRTValTypeName(self, rtval):
        """Returns the name of the type, handling extracting the name from KL RTVals.

        Args:
            rtval (rtval): The rtval to extract the name from.

        Returns:
            bool: True if successful.

        """

        if ks.isRTVal(rtval):
            return json.loads(rtval.type("Type").jsonDesc("String").getSimpleType())['name']
        else:
            return "None"

    # ==================
    # Config Methods
    # ==================

    def registerConfig(self, configClass):
        """Registers a config Python class with the KrakenSystem so ti can be built by the rig builder.

        Args:
            configClass (str): The Python class of the config

        """

        configModulePath = configClass.__module__ + "." + configClass.__name__
        if configModulePath in self.registeredConfigs:
            # we allow reregistring of configs because as a config's class is edited
            # it will be re-imported by python(in Maya), and the classes reregistered.
            pass

        self.registeredConfigs[configModulePath] = configClass


    def getConfigClass(self, className):
        """Returns the registered Python config class with the given name

        Args:
            className (str): The name of the Python config class

        Returns:
            object: The Python config class

        """

        if className not in self.registeredConfigs:
            raise Exception("Config with that class not registered:" + className)

        return self.registeredConfigs[className]


    def getConfigClassNames(self):
        """Returns the names of the registered Python config classes

        Returns:
            list: The array of config class names.

        """

        return self.registeredConfigs.keys()

    # ==================
    # Component Methods
    # ==================

    def registerComponent(self, componentClass):
        """Registers a component Python class with the KrakenSystem so ti can be built by the rig builder.

        Args:
            componentClass (str): The Python class of the component

        """
        componentClassPath = componentClass.__module__ + "." + componentClass.__name__
        if componentClassPath in self.registeredComponents:
            # we allow reregistring of components because as a component's class is edited
            # it will be re-imported by python(in Maya), and the classes reregistered.
            pass

        self.registeredComponents[componentClassPath] = componentClass


    def getComponentClass(self, className):
        """Returns the registered Python component class with the given name

        Args:
            className (str): The name of the Python component class

        Returns:
            object: The Python component class

        """

        if className not in self.registeredComponents:
            raise Exception("Component with that class not registered:" + className)

        return self.registeredComponents[className]


    def reloadAllComponents(self):
        """Force the reload of all registered component modules

        Return:
        True if successful.

        """

        for componentClassPath in self.registeredComponents:
            componentModulePath = self.registeredComponents[componentClassPath].__module__
            if componentModulePath in sys.modules:
                del(sys.modules[componentModulePath])

        prevRegsteredComponents = self.registeredComponents.copy()
        self.registeredComponents = {}

        for componentClassPath in prevRegsteredComponents:
            componentModulePath = prevRegsteredComponents[componentClassPath].__module__
            try:
                importlib.import_module(componentModulePath)
            except Exception as e:
                stack = traceback.format_tb(sys.exc_info()[2])
                exception_list = []
                exception_list.extend(stack)
                exception_list.extend(traceback.format_exception_only(sys.exc_info()[0], sys.exc_info()[1]))

                exception_str = "Traceback (most recent call last):\n"
                exception_str += "".join(exception_list)
                # Removing the last \n
                exception_str = exception_str[:-1]
                print(exception_str)

        return True


    def getComponentClassNames(self):
        """Returns the names of the registered Python component classes

        Returns:
            list: The array of component class names.

        """

        return self.registeredComponents.keys()


    def loadComponentModules(self):
        """Loads all the component modules and configs specified in the 'KRAKEN_PATHS' environment variable.

        The kraken_examples are loaded at all times.

        """


        def __importDirRecursive(path, parentModulePath=''):
            contents = os.listdir(path)
            moduleFilefound = False
            for item in contents:
                if os.path.isfile(os.path.join(path, item)):
                    if item == "__init__.py":
                        if parentModulePath == '':
                            modulePath = os.path.basename(path)

                            moduleParentFolder = os.path.split( path )[0]
                            if moduleParentFolder not in sys.path:
                                sys.path.append(moduleParentFolder)
                        else:
                            modulePath = parentModulePath + '.' + os.path.basename(path)
                        moduleFilefound = True


            if moduleFilefound:
                for item in contents:
                    if os.path.isfile(os.path.join(path, item)):
                        # parse all the files of given path and import python modules
                        if item.endswith(".py") and item != "__init__.py":
                            module = modulePath+"."+item[:-3]
                            try:
                                print "importlib.import_module(:" + str(module)
                                importlib.import_module(module)

                            except ImportError, e:
                                print e
                                for arg in e.args:
                                    print arg

                            except Exception, e:
                                for arg in e.args:
                                    print arg


            for item in contents:
                if os.path.isdir(os.path.join(path, item)):
                    if moduleFilefound:
                        __importDirRecursive(os.path.join(path, item), modulePath)
                    else:
                        __importDirRecursive(os.path.join(path, item))


        # find the kraken examples module in the same folder as the kraken module.
        examplePaths = os.path.join(os.path.dirname(os.path.dirname(kraken.__file__)), 'kraken_examples')
        __importDirRecursive(examplePaths)

        pathsVar = os.getenv('KRAKEN_PATHS')
        if pathsVar is not None:
            pathsList = pathsVar.split(';')
            for path in pathsList:

                if path == '':
                    continue

                if not os.path.exists(path):
                    print "Invalid Kraken Path: " + path
                    continue

                __importDirRecursive(path)


    @classmethod
    def getInstance(cls):
        """This class method returns the singleton instance for the KrakenSystem

        Returns:
            object: The singleton instance.

        """

        if cls.__instance is None:
            cls.__instance = KrakenSystem()

        return cls.__instance




ks = KrakenSystem.getInstance()
