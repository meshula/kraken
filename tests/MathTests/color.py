import json

from kraken.core.maths import *


color0 = Color()
color0.r = 1.0
print "color0:" + str(color0)
color1 = Color(1.0, 0.0, 0.74, 0.5)
print "color1:" + str(color1)
print "lerp:" + str(color0.linearInterpolate(color1, 0.5))

# Random color will never pass
# print "rand:" + str(Color.randomColor(0.5))

