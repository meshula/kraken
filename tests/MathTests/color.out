color0:Color(1.0,0.0,0.0,0.0)
color1:Color(1.0,0.0,0.0,0.0)
lerp:Color(1.0,0.0,0.0,0.0)
